include/benchmark/benchmark.h

```diff
@@ -1302,6 +1303,77 @@ struct SystemInfo {
   BENCHMARK_DISALLOW_COPY_AND_ASSIGN(SystemInfo);
 };
 
+// BenchmarkName wraps a std::string containing different fields
+// of a benchmark's name (e.g. the 'root' name, the benchmark's
+// 'args' etc.).
+class BenchmarkName {
+ public:
+  // Different fields of the benchmark's name.
+  enum Field {
+    // The name of the function being invoked in the benchmark.
+    kRoot = 1U,
+    // The arguments supplied to the benchmark, e.g. via Range().
+    kArgs = 1U << 1U,
+    // The minimum time specified for a benchmark.
+    kMinTime = 1U << 2U,
+    // The number of iterations specified for a benchmark.
+    kIterations = 1U << 3U,
+    // The number of repetitions specified for a benchmark.
+    kRepetitions = 1U << 4U,
+    // The type of time to measure for a benchmark.
+    kTimeType = 1U << 5U,
+    // The number of concurrent threads for a benchmark.
+    kThreads = 1U << 6U,
+    // All fields.
+    kAll =
+        kRoot | kArgs | kMinTime | kIterations | kRepetitions | kTimeType | kThreads
+  };
+
+  // Construct BenchmarkName with an 'empty' name.
+  BenchmarkName();
+
+  // Construct BenchmarkName with a given 'root'.
+  BenchmarkName(std::string root);
+
+  // Append 'field_content' to a given field of the name.
+  // Each appended element of content for a given field will be
+  // separated by a '/', and additionally each field will
+  // be separated by a '/'.
+  BenchmarkName& append(Field field, const std::string& field_content);
+
+  // Get the benchmark's name including all the fields set in 'fields'
+  std::string get(Field fields) const;
+
+  // Get the full name of the benchmark including all fields.
+  const std::string& str() const { return name_; }
+
+  // Get the size of the full name of the benchmark.
+  size_t size() const { return name_.size(); }
+
+ private:
+  bool is_set(Field field, std::size_t enumerator_index) const;
+  size_t start_offset(std::size_t enumerator_index) const;
+  size_t end_offset(std::size_t enumerator_index) const;
+
+  static const char separator = '/';
+  static const std::size_t enumerator_count = 7;
```

Comment:
this seems fragile. it might be worth allowing the enums to take their natural values and using `kAll` (which would have value 7) as the enumerator count. Then you'd just need to shift wherever you access the field in the implementation.

Reply:
I agree this is fragile. With modern c++ we could compute at compile time the number of set bits in `kAll`, which would do the trick, but it has to be done in the header so wouldn't work for <= c++03 clients.

I think your suggestion would make it tricky to chose a subset of name fields - composing the bits of enumerator values seems natural for that.

Of course, as you point out, only `kAll` and `kAll | ~kArgs` are currently used, so we could just add an additional enumerator for 'name without the args' request (or, as you suggest below, drop the enumeration entirely and just have two accessors for the currently-used cases).

Reply:
Does this have to be in the public header?
I think this is an internal functionality.

Reply:
It's a member of `BenchmarkReporter`, which is the only reason it's in the header. I could store a pointer to it in `BenchmarkReporter` so that it only requires a forward-declaration?